packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -379,6 +379,34 @@ class DataSourceWithBackend<
       });
     });
   }
+
+  /**
+   * EXPERIMENTAL. Calls migration endpoint. Requires grafanaAPIServerWithExperimentalAPIs or datasourceAPIServers feature toggle.
+   */
+  async postMigrateQuery(query: TQuery): Promise<TQuery> {
```

Comment:
why would we only want to do one, and not the set that are passed into the datasource?

Ideally we can find a way to call this once when the dashboard is loaded 🤔

Reply:
Somethign like https://github.com/grafana/grafana/compare/sanitize-queries maybe?

Reply:
How about `QueryEditorRow` then? If the plugin has defined a migration, I would expect the framework to make sure it's executed at the right time and not the responsibility if the plugin author. If not possible to do this in `QueryEditorRow`, it may be worth putting in some effort to harmonize the rendering of query editors across the frontend so that we have more control of it.

Reply:
Checking `QueryEditorRow`, it's used by both dashboards and alerts, are we confident that no other thing will render a `QueryEditor` directly? From the top of my mind, plugins may directly use the same `QueryEditor` as the template variable editor and I don't think we can do much for that case.

It'd be indeed nicer if the user don't need to handle the migration themselves (but not a crazy thing to ask since they will need to manually handle that anyway in the backend).

In any case, `QueryEditorRow` handles one query at a time so we couldn't still run multiple migrations at once (I think this is fine).

Reply:
>are we confident that no other thing will render a QueryEditor directly? 

IDK, but feels like we have the same problem with for example [getDefaultQuery](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/datasource.ts/#L376) then. Again, feels like we would benefit from taking control of the QueryEditor rendering.

How about adding a HOC (or a similar kind of abstraction that wraps the query editor) upon registration [here](https://github.com/grafana/grafana/blob/405887eebf024aa503d68440c85ca1437e87a23c/packages/grafana-data/src/types/datasource.ts#L69)? Then this HOC would call migrateQuery (if implemented) and render a skeleton meanwhile migrating. That way, the query editor is guaranteed to always receive the new, migrated props (which I believe is not the case in the example above?).

Reply:
> How about adding a HOC (or a similar kind of abstraction that wraps the query editor) upon registration [here](https://github.com/grafana/grafana/blob/405887eebf024aa503d68440c85ca1437e87a23c/packages/grafana-data/src/types/datasource.ts#L69)?

Probably not the best option either, those are React components so adding there a function hook may be confusing. Also, once the hook is registered (`setMigrationHandler` for example), something needs to call it so we are back to the same problem since we don't have a hook that is always executed before rendering the query editor.

From a quick search, I found these components, that render directly a `QueryEditor` and therefore, would be out if we just cover `QueryEditorRow`:

 - https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/alerting/unified/components/rule-editor/ExpressionEditor.tsx#L67
 -  https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/alerting/unified/components/rule-editor/RecordingRuleEditor.tsx/#L99
 - https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/annotations/components/StandardAnnotationQueryEditor.tsx/#L206
 - https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/correlations/Forms/QueryEditorField.tsx#L28
 - https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/variables/editor/getVariableQueryEditor.tsx#L36

So if we want to modify how we render the query editor we would need to change at least all those components (and maybe there are more).

I don't think we _need_ to have the infrastructure to have a "prerender" hook for this feature, at least for a first version. If in the future we have it, then we know we can execute things there like `getDefaultQuery` or `migrateQuery` and people would be able to simplify their QueryEditors but that's something that doesn't necessarily needs to block this extension.

Reply:
> Probably not the best option either, those are React components so adding there a function hook may be confusing. Also, once the hook is registered (`setMigrationHandler` for example), something needs to call it so we are back to the same problem since we don't have a hook that is always executed before rendering the query editor.

I'm not suggesting a hook, but a HOC. So you'd change the `setQueryEditor` method in datasource.ts to something along those lines (pseudo code). 
```ts
setQueryEditor(QueryEditor: ComponentType<QueryEditorProps<DSType, TQuery, TOptions>>) {   
  this.components.QueryEditor = QueryEditorWithExtra(QueryEditor);
  return this;
}
// Wrapped query editor
function QueryEditorWithExtra<T>(QueryEditor: ComponentType<QueryEditorProps<T>) {
  const WithExtra = (props: QueryEditorProps<T>) => {
    useEffect(() => {
      // call datasource.migrateQuery, save state and pass it to the actual editor
    }, []);
    return <QueryEditor {...props} />;
  };
  return WithExtra;
}

```

Reply:
I misunderstood you, that actually works! I have updated the PR, let me know what you think.